### PR TITLE
[Fix] Refactor Deformable and ProjAttn for MvP

### PR DESCRIPTION
**Pull request by wqyin**
1. Separate the `ProjAttn` class and relevant functions from `Deformable` package.
2. Add `ProjAttn` to the model registry.
3. Simplify the download and installation process of the `Deformable` package.
4. Update benchmarks and tutorials.

### CODE REVIEW

File: xrmocap/model/architecture/multi_view_pose_transformer.py [hunk 1]
```diff
@@ -51,12 +52,12 @@ def __init__(self, is_train: bool, logger: Union[None, str,
...
-                 pred_class_fuse: str, projattn_pose_embed_mode: str,
+                 pred_class_fuse: str, projattn_pos_embed_mode: str,
                  query_adaptation: bool, convert_kp_format_indexes: list,
-                 backbone_setup: dict, decoder_layer_setup: dict,
-                 decoder_setup: dict, pos_encoding_setup: dict,
-                 pose_embed_setup: dict, matcher_setup: dict,
-                 criterion_setup: dict, space_size: list,
+                 backbone_setup: dict, proj_attn_setup: dict,
+                 decoder_layer_setup: dict, decoder_setup: dict,
+                 pos_encoding_setup: dict, pose_embed_setup: dict,
```
Comment (jiaqiAA) [inline xrmocap/model/architecture/multi_view_pose_transformer.py:59]: I want to make sure, should it be `pose_embed_setup` or `pos_embed_setup`?

File: xrmocap/model/mvp/projattn.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                 n_points=4,
+                 projattn_pos_embed_mode='use_rayconv'):
+        """Projective Attention Module Modified from DeformableConvV2 and
+        Deformable DETR https://github.com/chengdazhi/Deformable-
+        Convolution-V2-PyTorch/tree/pytorch_1.0.0
+        https://github.com/fundamentalvision/Deformable-DETR.
+
+        Args:
+            d_model (int, optional):
+                hidden dimension. Defaults to 256.
```
Comment (jiaqiAA) [inline xrmocap/model/mvp/projattn.py:31]: Capitalize the first letter?

File: xrmocap/model/mvp/projattn.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        elif self.projattn_pos_embed_mode == 'ablation_not_use_rayconv':
+            for lvl in range(feat_lvls):
+                ref_point_feat_views_alllvs.append(
+                    F.grid_sample(src_views[lvl],
+                                  sample_grid[:, :, lvl:lvl +
+                                              1, :]).squeeze(-1).permute(
+                                                  0, 2, 1))
+            input_flatten = torch.cat([src.flatten(2) for src in src_views],
+                                      dim=-1).permute(0, 2, 1)
+
```
Comment (jiaqiAA) [inline xrmocap/model/mvp/projattn.py:181]: Shall we get an error when `self.projattn_pos_embed_mode` is another type?

File: xrmocap/model/architecture/multi_view_pose_transformer.py (hunk 1 above)
Comment (wqyin) [reply xrmocap/model/architecture/multi_view_pose_transformer.py:54]: `pose_embed` is the embedding layers for keypoints.
`pos_embed` is the positional embedding mode for the transformer.

File: xrmocap/model/mvp/projattn.py (hunk 2 above)
Comment (wqyin) [reply xrmocap/model/mvp/projattn.py]: Fixed in [ad43f3b](https://github.com/openxrlab/xrmocap/pull/21/commits/ad43f3b61eb55fb3c7998d07ae3e7b87c871347f).

File: xrmocap/model/mvp/projattn.py (hunk 3 above)
Comment (wqyin) [reply xrmocap/model/mvp/projattn.py:196]: Added in [ad43f3b](https://github.com/openxrlab/xrmocap/pull/21/commits/ad43f3b61eb55fb3c7998d07ae3e7b87c871347f), logger will log the error if unrecognized mode is given.

File: xrmocap/model/mvp/projattn.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            None, None, \
+            grad_sampling_loc, \
+            grad_attn_weight, None
+
+
+def _is_power_of_2(n):
+    if (not isinstance(n, int)) or (n < 0):
+        raise ValueError('invalid input '
+                         'for _is_power_of_2: '
+                         '{} (type: {})'.format(n, type(n)))
```
Comment (yl-1993) [inline xrmocap/model/mvp/projattn.py:280]: f-string?

File: xrmocap/model/mvp/projattn.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                camera_ray_embeds,
+                input_spatial_shapes,
+                input_level_start_index,
+                input_padding_mask=None):
+        """Forward function for Projective Attention.
+
+        Args:
+            query :
+                (n_views, Length_{query}, C)
+            reference_points :
```
Comment (yl-1993) [inline xrmocap/model/mvp/projattn.py:135]: Describe the semantic meaning of`reference_points` and `query`?

File: xrmocap/model/mvp/projattn.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+            n_levels (int, optional):
+                Number of feature levels. Defaults to 4.
+            n_heads (int, optional):
+                Number of attention heads. Defaults to 8.
+            n_points (int, optional):
+                Number of sampling points per attention head
+                per feature level. Defaults to 4.
+            projattn_pos_embed_mode (str, optional):
+                The positional embedding mode of projective
+                attention. Defaults to 'use_rayconv'.
```
Comment (yl-1993) [inline xrmocap/model/mvp/projattn.py:53]: Write out all possible values? e.g., The mode includes 'use_rayconv' and  'use_2d_coordconv'. The former denotes xxx, while the latter means xxx. Defaults to  'use_rayconv'.

File: xrmocap/model/mvp/projattn.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        # 2 which is more efficient in our CUDA implementation
+        if not _is_power_of_2(_d_per_head):
+            self.logger.warning("You'd better set d_model in Deform to "
+                                'make the dimension of each attention '
+                                'head a power of 2 which is more efficient '
+                                'in our CUDA implementation.')
+            warnings.warn("You'd better set d_model in Deform to "
+                          'make the dimension of each attention '
+                          'head a power of 2 which is more efficient '
+                          'in our CUDA implementation.')
```
Comment (yl-1993) [inline xrmocap/model/mvp/projattn.py:76]: Why we need `warnings` and `logger.warning` at the same time?

File: scripts/download_install_deformable.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+mkdir xrmocap_download
+cd xrmocap_download
+wget -q https://openxrlab-share.oss-cn-hongkong.aliyuncs.com/xrmocap/Deformable.tar
```
Comment (yl-1993) [inline scripts/download_install_deformable.sh:3]: After checking out this tar, I think we may maintain the cuda code directly in our codebase, which would be easier to track if anything from Deformable is updated.

We can leave it for another PR.

File: xrmocap/model/mvp/projattn.py (hunk 7 above)
Comment (wqyin) [reply xrmocap/model/mvp/projattn.py]: Fixed in [d5025b7](https://github.com/openxrlab/xrmocap/pull/21/commits/d5025b700c504f02e48cdaae421d19a4d3d2c8d4), `warnings` removed.

File: xrmocap/model/mvp/projattn.py (hunk 6 above)
Comment (wqyin) [reply xrmocap/model/mvp/projattn.py]: Added in [d5025b7](https://github.com/openxrlab/xrmocap/pull/21/commits/d5025b700c504f02e48cdaae421d19a4d3d2c8d4).

File: xrmocap/model/mvp/projattn.py (hunk 5 above)
Comment (wqyin) [reply xrmocap/model/mvp/projattn.py:139]: Added in [d5025b7](https://github.com/openxrlab/xrmocap/pull/21/commits/d5025b700c504f02e48cdaae421d19a4d3d2c8d4).

File: xrmocap/model/mvp/projattn.py (hunk 4 above)
Comment (wqyin) [reply xrmocap/model/mvp/projattn.py]: Fixed in [d5025b7](https://github.com/openxrlab/xrmocap/pull/21/commits/d5025b700c504f02e48cdaae421d19a4d3d2c8d4).

File: xrmocap/model/mvp/projattn.py [hunk 9]
```diff
@@ -50,7 +49,14 @@ def __init__(self,
                 per feature level. Defaults to 4.
             projattn_pos_embed_mode (str, optional):
                 The positional embedding mode of projective
-                attention. Defaults to 'use_rayconv'.
+                attention. The mode includes 'use_rayconv',
+                'use_2d_coordconv' and 'ablation_not_use_rayconv'.
+                'use_raycon' uses camera ray dirction asd the positional
```
Comment (yl-1993) [inline xrmocap/model/mvp/projattn.py:54]: asd -> as?

Comment (wqyin) [reply xrmocap/model/mvp/projattn.py]: Fixed in [e47f350](https://github.com/openxrlab/xrmocap/pull/21/commits/e47f35020fa26325b63a01427876cffed8cae452).

**Review by yl-1993 — APPROVED** [review]
